File: mod/common/java/xen/mod/core/Blocks.java
package xen.mod.core;

/** Block categories Xen perceives (must match xen/blocks.py). */
public final class Blocks {
	public static final int AIR = 0, GRASS = 1, DIRT = 2, STONE = 3, LOG = 4, LEAVES = 5, COAL = 6, IRON = 7,
			GOLD = 8, DIAMOND = 9, LAVA = 10, WATER = 11, BEDROCK = 12, COUNT = 13;

	public static final String[] NAMES = {"air", "grass", "dirt", "stone", "log", "leaves", "coal ore", "iron ore",
			"gold ore", "diamond ore", "lava", "water", "bedrock"};
	public static final boolean[] SOLID = {false, true, true, true, true, true, true, true, true, true, false, false, true};
	public static final boolean[] OPAQUE = {false, true, true, true, true, true, true, true, true, true, true, false, true};
	/** Satisfaction of mining one (value of what drops). */
	public static final float[] VALUE = {0f, 0.05f, 0.05f, 0.1f, 1f, 0f, 1.5f, 3f, 4f, 10f, 0f, 0f, 0f};
	public static final boolean[] TREASURE = new boolean[COUNT];
	public static final float[] VALUE_NORM = new float[COUNT];

	static {
		float max = 0;
		for (float v : VALUE) max = Math.max(max, v);
		for (int i = 0; i < COUNT; i++) {
			TREASURE[i] = VALUE[i] >= 1f;
			VALUE_NORM[i] = (float) (Math.log1p(VALUE[i]) / Math.log1p(max));
		}
	}

	private Blocks() {}
}

File: xen/blocks.py
"""Block and item catalogue shared by every world Xen can live in.

Real Minecraft has hundreds of blocks; Xen perceives them through a small set
of categories that matter for survival and progress.
"""
import numpy as np

(AIR, GRASS, DIRT, STONE, LOG, LEAVES, COAL, IRON, GOLD, DIAMOND,
 LAVA, WATER, BEDROCK) = range(13)
NUM_BLOCKS = 13

NAMES = ("air", "grass", "dirt", "stone", "log", "leaves", "coal_ore",
         "iron_ore", "gold_ore", "diamond_ore", "lava", "water", "bedrock")

SOLID = np.array([0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1], dtype=bool)
LIQUID = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0], dtype=bool)

# Ticks of holding MINE needed to break a block; -1 means it cannot be mined.
HARDNESS = np.array([-1, 2, 2, 4, 3, 1, 5, 6, 6, 8, -1, -1, -1])

# What ends up in the inventory when a block breaks.
DROPS = {GRASS: "dirt", DIRT: "dirt", STONE: "cobblestone", LOG: "log",
         COAL: "coal", IRON: "raw_iron", GOLD: "raw_gold", DIAMOND: "diamond"}

# How satisfying it is to obtain one of each item.
ITEM_VALUE = {"dirt": 0.05, "cobblestone": 0.1, "log": 1.0, "coal": 1.5,
              "raw_iron": 3.0, "raw_gold": 4.0, "diamond": 10.0, "food": 0.3}

# Common items lose their appeal as Xen piles them up (a stack of dirt is plenty);
# ores, logs and food never do.
COMMON = ("dirt", "cobblestone")


def satisfaction(item, carried):
    """How rewarding it is to get one more `item` when already carrying `carried` of it."""
    value = ITEM_VALUE[item]
    return value * 16.0 / (16.0 + carried) if item in COMMON else value


# Items that can be placed back into the world, and the block they become.
PLACEABLE = {"cobblestone": STONE, "dirt": DIRT}

INVENTORY_ITEMS = ("dirt", "cobblestone", "log", "coal", "raw_iron",
                   "raw_gold", "diamond", "food")

VALUE = np.zeros(NUM_BLOCKS, dtype=np.float32)
for _block, _item in DROPS.items():
    VALUE[_block] = ITEM_VALUE[_item]

# Blocks worth walking towards.
TREASURE = VALUE >= 1.0
